torch/optim/lbfgs.py

```diff
@@ -3,8 +3,165 @@
 from .optimizer import Optimizer
 
 
+def _cubic_interpolate(x1, f1, g1, x2, f2, g2, bounds=None):
+    # ported from https://github.com/torch/optim/blob/master/polyinterp.lua
+    # Compute bounds of interpolation area
+    if bounds is not None:
+        xmin_bound, xmax_bound = bounds
+    else:
+        xmin_bound, xmax_bound = (x1, x2) if x1 <= x2 else (x2, x1)
+
+    # Code for most common case: cubic interpolation of 2 points
+    #   w/ function and derivative values for both
+    # Solution in this case (where x2 is the farthest point):
+    #   d1 = g1 + g2 - 3*(f1-f2)/(x1-x2);
+    #   d2 = sqrt(d1^2 - g1*g2);
+    #   min_pos = x2 - (x2 - x1)*((g2 + d2 - d1)/(g2 - g1 + 2*d2));
+    #   t_new = min(max(min_pos,xmin_bound),xmax_bound);
+    d1 = g1 + g2 - 3 * (f1 - f2) / (x1 - x2)
+    d2_square = d1 ** 2 - g1 * g2
+    if d2_square >= 0:
+        d2 = d2_square.sqrt()
+        if x1 <= x2:
+            min_pos = x2 - (x2 - x1) * ((g2 + d2 - d1) / (g2 - g1 + 2 * d2))
+        else:
+            min_pos = x1 - (x1 - x2) * ((g1 + d2 - d1) / (g1 - g2 + 2 * d2))
+        return min(max(min_pos, xmin_bound), xmax_bound)
```

Comment:
As mentioned before equation 3.59, the minimum of the cubic function will occur either at min_pos, xmin_bound, or xmax_bound. Since equation 3.59 doesn’t provide the value of the cubic polynomial at the minimum, and computing it might increase the computational cost, we could simply return xmin_bound or xmax_bound depending on which one has a lower value of f, whenever min_pos falls outside the range of xmin_bound and xmax_bound.

An example (which could be used for testing this algorithm) would be interpolating the function f(x)=-x(x-2)**2 on [0,1]. The interpolation is exact, since it is a cubic polynomial. The values are f(0)=0, f(1)=1, g(0)=f'(0)=4, g(1)=f'(1)=5. The local minimum would be found at x=2 outside the range [0,1]. This algorithm would return x=1 since that is the closest point to the range, but the minimum on this range is at 0.

That being said, this is indeed implementing what is done in https://github.com/torch/optim/blob/master/polyinterp.lua#L131, so I am okay with using either as authoritative to move ahead, and leaving this as a later improvement.

Reply:
I agree that it might be better to compare f1, f2 and pick x_i with smaller f_i when min_pos is outside of the interval. But I would like to stick with original implementations, and leave this for later improvement.

> Does minFunc also do as in lua?

Yes, minFunc also does so.